components/zeekopenbsm/src/openbsmconsumer.cpp

```diff
@@ -0,0 +1,363 @@
+    } else
+      event.header.success = 0;
+    break;
+  }
+  }
+}
+void OpenbsmConsumer::extractSocketInet(IOpenbsmConsumer::Event &event,
+                                        tokenstr_t tok) {
+
+  switch (tok.id) {
+  case AUT_SOCKINET32: {
+    if (event.type == Event::Type::Bind) {
+      event.header.remote_address = "0.0.0.0";
+      event.header.remote_port = 0;
+      event.header.local_address = getIpFromToken(tok);
+      event.header.local_port = ntohs(tok.tt.sockinet_ex32.port);
+    } else {
+      event.header.remote_address = getIpFromToken(tok);
+      event.header.remote_port = ntohs(tok.tt.sockinet_ex32.port);
+      event.header.local_address = "0.0.0.0";
+      event.header.local_port = 0;
+    }
+    if (tok.tt.sockinet_ex32.family == 2) {
+      event.header.family = 2;
+    } else if (tok.tt.sockinet_ex32.family == 26) {
+      event.header.family = 10;
+    } else {
+      event.header.family = 0;
+    }
+
+    break;
+  }
+  case AUT_SOCKINET128: {
+    if (event.type == Event::Type::Bind) {
+      event.header.remote_address = "0.0.0.0";
+      event.header.remote_port = 0;
+      event.header.local_address = getIpFromToken(tok);
+      event.header.local_port =
+          static_cast<std::int64_t>(ntohs(tok.tt.sockinet_ex32.port));
+    } else {
+      event.header.remote_address = getIpFromToken(tok);
+      event.header.remote_port =
+          static_cast<std::int64_t>(ntohs(tok.tt.sockinet_ex32.port));
+      event.header.local_address = "0.0.0.0";
+      event.header.local_port = 0;
+    }
+    if (tok.tt.sockinet_ex32.family == 2) {
+      event.header.family = 2;
+    } else if (tok.tt.sockinet_ex32.family == 26) {
+      event.header.family = 10;
+    } else {
+      event.header.family = 0;
+    }
+    break;
+  }
+  }
+}
+
+Status
+OpenbsmConsumer::populateSocketEvent(Event &event,
+                                     const std::vector<tokenstr_t> &tokens) {
+
+  for (const auto &tok : tokens) {
+    if (tok.id == AUT_SOCKUNIX) {
+      // we filter unix sockets out
+      continue;
+    }
+    switch (tok.id) {
+    case AUT_HEADER32:
+    case AUT_HEADER64:
+    case AUT_HEADER32_EX:
+    case AUT_HEADER64_EX:
+      extractHeader(event, tok);
+      break;
+    case AUT_SUBJECT32:
+    case AUT_SUBJECT64:
+    case AUT_SUBJECT32_EX:
+    case AUT_SUBJECT64_EX:
+      extractSubject(event, tok);
+      break;
+    case AUT_RETURN32:
+    case AUT_RETURN64:
+      extractReturn(event, tok);
+      break;
+    case AUT_SOCKINET32:
+    case AUT_SOCKINET128:
+      extractSocketInet(event, tok);
+      break;
+    }
+  } // for loop
+
+  return Status::success();
+}
+
+Status OpenbsmConsumer::parseMessage() {
+
+  Status status;
+  Event event;
+
+  u_char *buffer = nullptr;
+
+  int reclen = au_read_rec(audit_pipe, &buffer);
+
+  if (reclen <= 0) {
+    return Status::failure("Could not openbsm fetch message");
+  }
+
+  tokenstr_t tok;
+  std::vector<tokenstr_t> tokens{};
+  tokens.reserve(12);
```

Comment:
Where is this number coming from? Shouldn't this be (derived from) `reclen` instead?

Reply:
Basically, we need an upper bound on the number of token strings that can appear per socket bind/connect event. I could not find any documentation that can give me an upper bound value. Further, I don't think we can derive that number from `reclen` because `reclen` represents total bytes in all the token strings per event rather than the total number of token strings per event. So, I looked into osquery code and they were using 12 as an upper bound.

Reply:
Let's either add a comment capturing how we arrive at this number, or get rid of the `reserve`.

Reply:
I'll get rid of the `reserve` because I am not sure how memory saving it buys us.